clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -508,17 +511,16 @@ private static int parseAcks(String acksString) {
      * Wait for cluster metadata including partitions for the given topic to be available.
      * @param topic The topic we want metadata for
      * @param maxWaitMs The maximum time in ms for waiting on the metadata
-     * @return The amount of time we waited in ms
+     * @param beginTimeMs Start time from which maxWaitMs applies
+     * @return The cluster containing topic metadata
      */
-    private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
+    private Cluster waitOnMetadata(String topic, long maxWaitMs, long beginTimeMs) throws InterruptedException {
         // add topic to metadata topic list if it is not there already and reset expiry
         this.metadata.add(topic);
-        if (metadata.fetch().partitionsForTopic(topic) != null)
-            return 0;
```

Comment:
@ijuma Oops, thank you for pointing that out. I have updated the code to avoid the unnecessary `time.milliseconds()`. The code uses an AtomicLong to pass-by-reference (there is another similar usage in the same file), but I dont mind adding a class to return both cluster and wait time if that is preferred.